src/nvapi_gpu.cpp

```diff
@@ -348,4 +348,95 @@ extern "C" {
                 return Error(str::format(n, ": ", adapter->NvmlErrorString(result)));
         }
     }
+
+    NvAPI_Status __cdecl NvAPI_GPU_GetAllClockFrequencies(NvPhysicalGpuHandle hPhysicalGpu, NV_GPU_CLOCK_FREQUENCIES *pClkFreqs) {
+        constexpr auto n = "NvAPI_GPU_GetAllClockFrequencies";
+        static bool alreadyLoggedNoNvml = false;
+        static bool alreadyLoggedHandleInvalidated = false;
+        static bool alreadyLoggedOk = false;
+
+        if (nvapiAdapterRegistry == nullptr)
+            return ApiNotInitialized(n);
+
+        if (pClkFreqs == nullptr)
+            return InvalidArgument(n);
+
+        if (pClkFreqs->version != NV_GPU_CLOCK_FREQUENCIES_VER_1 && pClkFreqs->version != NV_GPU_CLOCK_FREQUENCIES_VER_2 && pClkFreqs->version != NV_GPU_CLOCK_FREQUENCIES_VER_3)
+            return IncompatibleStructVersion(n);
+
+        // Only check for CURRENT_FREQ, and not for the other types ie. BOOST or DEFAULT for now
+        if (pClkFreqs->ClockType != NV_GPU_CLOCK_FREQUENCIES_CURRENT_FREQ)
+            return NotSupported(n);
+
+        auto adapter = reinterpret_cast<NvapiAdapter*>(hPhysicalGpu);
+        if (!nvapiAdapterRegistry->IsAdapter(adapter))
+            return ExpectedPhysicalGpuHandle(n);
+
+        if (!adapter->HasNvml())
+            return NoImplementation(str::format(n, ": NVML not loaded"), alreadyLoggedNoNvml);
+
+        if (!adapter->HasNvmlDevice())
+            return HandleInvalidated(str::format(n, ": NVML available but current adapter is not NVML compatible"), alreadyLoggedHandleInvalidated);
+
+        // Reset all clock data for all domains
+        for (auto i = 4U; i < NVAPI_MAX_GPU_PUBLIC_CLOCKS; i++){
+            pClkFreqs->domain[i].frequency = 0;
+            pClkFreqs->domain[i].bIsPresent = FALSE;
+        }
+
+        unsigned int clock;
+        // Seemingly we need to do nvml call on a "per clock unit" to get the clock
+        // Set the availability of the clock to TRUE and the nvml read clock in the nvapi struct
+        auto resultgpu = adapter->NvmlDeviceGetClockInfo(NVML_CLOCK_GRAPHICS, &clock);
+            switch (resultgpu) {
+                case NVML_SUCCESS:
+                    pClkFreqs->domain[NVAPI_GPU_PUBLIC_CLOCK_GRAPHICS].bIsPresent = TRUE;
+                    pClkFreqs->domain[NVAPI_GPU_PUBLIC_CLOCK_GRAPHICS].frequency = static_cast<long unsigned int>(clock * 1000);
```

Comment:
Do we really need a `static_cast` here?

Reply:
I think we should be able to skip the cast, I believe that the type of `frequency` (which is `NvU32`) is the same as the return type of `nvmlDeviceGetClockInfo` (which is `unsigned int`) on every platform that matters (and on 64–bit Linux, casting to `unsigned long` would actually make it too large to fit in there).

Reply:
Yes, I also think this is not needed. Usually I'm lazy and just follow what CLion (which uses clang-tidy underneath) warns me about, but I haven't looked at this branch in the IDE yet...